Demo/kobobooks parser

LukaszCzubernat opened the pull request:
Yo!
Finally I had time to finish and MERGE (git hates me :sob: ) with master my parse.
My aim was to get books from https://store.kobobooks.com/p/free-ebooks
I haven't yet add tests, they will appear in near future 

// for more details read commit comment of https://github.com/team-javapocalypse-ja/OgreRobot/commit/7b070d750361d68e6c30d4e6119d756c2c997123

Commit:
KoboBooks parser
I've create prototype parser for kobobooks.com. Parser get almost all data about books that are required (except tags)
This implementation need few refactors to be done and more optimalization, bacause parsing from that site have high time complexity.

Commit:
Refactor of logic
I've fixed and extended logic of parser for kobobooks.com page. I still need to get rid of msvm in KoboBooksParser but I'll do this as soon as I rebese master on this branch. Then I'll add some tests for it. I know I broke TDD manner but in this case developing in TDD manner will be to hard and time consuming. Most of a time I spend on debugging and finding out how can I get date from html.

Commit:
Rebased master on branch and little refactor
I've rebased maseter onto this branch, resolved conflicts and did little refactor.

Commit:
Fix previous commits
By mistake I did merged branches properly. This commit fixes it

egolesor commented on RobotCrawler/pom.xml:
```diff
@@ -11,6 +11,9 @@
 
+    <properties>
+        <jsoup.version>1.9.2</jsoup.version>
+    </properties>
```
We already have it on main pom

LukaszCzubernat replied:
Removed

egolesor commented on RobotCrawler/src/main/java/parser/kobobooks/KoboBooksCategoryParser.java:
```diff
@@ -0,0 +1,37 @@
+                    .library(dataCollector.libraryFrom(document))
+                    .url(dataCollector.urlFrom(document)).build());
+        }
+        return books;
```
You could use BookDataFactory

LukaszCzubernat replied:
It will be hard to do this util we integrate all parsers

egolesor commented on RobotCrawler/src/main/java/parser/kobobooks/KoboPageDownloader.java:
```diff
@@ -0,0 +1,36 @@
+        Document document = null;
+        try {
+            if(Objects.isNull(login)) {login();}
+            document = Jsoup.connect(baseUrl.concat(additionalUrl))
```
We have something like DocumentBuilder. checkout if it can be useful in this place.

LukaszCzubernat replied:
I've removed this class and now I'm using DocumentBuilder.
Also I've added some additional behavior to DocumentBuilder

egolesor commented on RobotCrawler/src/main/java/parser/kobobooks/KoboPageDownloader.java:
```diff
@@ -0,0 +1,36 @@
+                .userAgent("Mozilla")
+                .method(Connection.Method.POST)
+                .execute();
+    }
```
:+1:

sledzgrzegorz commented on RobotCrawler/src/main/java/parser/kobobooks/KoboBookDataCollector.java:
```diff
@@ -0,0 +1,59 @@
+                .orElse(null);
+    }
+
+}
```
Method are small. I like this code :+1:

sledzgrzegorz commented on RobotCrawler/src/main/java/parser/kobobooks/KoboBooksParser.java:
```diff
@@ -0,0 +1,32 @@
+    private static final String freeBooksUrl = "/p/free-ebooks";
+    KoboBooksPageParser booksPageParser;
+    KoboBooksCategoryPageParser booksCategoryPageParser;
+    KoboBooksCategoryParser booksCategoryParser;
```
Maybe KoboBooksPageParser can be renamed to KoboBooksCategoriesUrlParser.
KoboBooksCategoryPageParser can be renamed to KoboBooksUrlParser.
KoboBooksCategoryParser can be renamed to KoboBooksParser.
But I have not idea how to change name of KoboBooksParser.

This is only suggestion :)

Commit:
Refactor after code review
I've refactored all classes and connected them to common 'downloader'. It's first step into ingegration with rest of parsers.
Also I've added looging to all classes - 'log them all'

Commit:
Second step in integrating components
I've refactored once again whole kobobooks parser package. Now it has more things in common with others parsers

Commit:
Restoring previous code
By mistake I've deleted files that should not be deleted and commited those changes

sledzgrzegorz commented on RobotCrawler/src/main/java/parser/DocumentBuilder.java:
```diff
@@ -2,26 +2,49 @@
 
     public Document buildFromFile() throws IOException {
+        log.debug("Parsing file ".concat(file.getAbsolutePath()));
         return Jsoup.parse(file, null);
```
Here you can pass Charset instead of null :)

Commit:
Code format and optimize import
I went through all classes and I've used format and optimize imports on them. Nothing fancy :smile:

sledzgrzegorz commented on RobotCrawler/src/main/java/parser/empik/PageDownloader.java:
```diff
@@ -0,0 +1,22 @@
+        Document document;
+        try {
+            document = Jsoup.connect(url).get();
+        } catch (IOException e) {
```
Maybe some information should be log if exception happened.

LukaszCzubernat replied:
Done

sledzgrzegorz commented on RobotCrawler/src/main/java/parser/DocumentBuilder.java:
```diff
@@ -27,14 +27,19 @@ public Document buildFromFile() throws IOException {
+                    .cookies(login.cookies())
+                    .get();
+        } catch (IOException e) {
+            log.error("");
```
error without message? :D

LukaszCzubernat replied:
Message added

sledzgrzegorz commented on RobotCrawler/src/test/java/parser/ebookscom/EBooksParserTest.java:
```diff
@@ -73,18 +74,12 @@ public void checkIfTheParsedDataCouldConvertToBookData(){
-        List<Element> offersUrl = parser.parse(documentFromUrl);
+        List<Element> offersUrl = parser.parse(documentOptional.get());
         List<BookData> bookDataList = BookDataFactory.newListBookData(offersUrl, getter, "art");
         System.out.println(bookDataList);
```
println should be removed :)

sledzgrzegorz commented on RobotCrawler/src/main/java/parser/kobobooks/KoboBooksParser.java:
```diff
@@ -17,25 +16,28 @@
             log.error("Error during parsing ".concat(e.getMessage()));
+        } catch (Exception e) {
+
         }
```
catch (Exception e) This also should be logged

sledzgrzegorz commented:
Ok, I wrote some comments. This code looks good, it can be merged :)

Commit:
Fixes after code review
I've changed few things after code review

LukaszCzubernat commented:
Ok, I will merge it but again I can't do rebase :cry:
So I'll close this pull request, delete remote branch and create new remote from existing local. Than I'll create PR.
